src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java

```diff
@@ -0,0 +1,109 @@
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.streaming.StreamingContext;
+import co.cask.cdap.etl.api.streaming.StreamingSource;
+import co.cask.gcp.common.GCPReferenceSourceConfig;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.storage.StorageLevel;
+import org.apache.spark.streaming.api.java.JavaDStream;
+import org.apache.spark.streaming.api.java.JavaReceiverInputDStream;
+import org.apache.spark.streaming.pubsub.PubsubUtils;
+import org.apache.spark.streaming.pubsub.SparkGCPCredentials;
+import org.apache.spark.streaming.pubsub.SparkGCPCredentials.Builder;
+import org.apache.spark.streaming.pubsub.SparkPubsubMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+/**
+ * Realtime source plugin to read from Google PubSub.
+ */
+@Plugin(type = StreamingSource.PLUGIN_TYPE)
+@Name("GoogleSubscriber")
+@Description("Streaming Source to read messages from Google PubSub.")
+public class GoogleSubscriber extends StreamingSource<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(GoogleSubscriber.class);
+  private static final Schema DEFAULT_SCHEMA =
+    Schema.recordOf("event",
+                    Schema.Field.of("message", Schema.of(Schema.Type.STRING)),
+                    Schema.Field.of("id", Schema.of(Schema.Type.STRING)),
+                    Schema.Field.of("timestamp", Schema.of(Schema.Type.STRING))
+    );
+
+  private SubscriberConfig config;
+
+  public GoogleSubscriber(SubscriberConfig conf) {
+    this.config = conf;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    super.configurePipeline(pipelineConfigurer);
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_SCHEMA);
+  }
+
+  @Override
+  public JavaDStream<StructuredRecord> getStream(StreamingContext streamingContext) {
+    String serviceAccountFilePath = config.getServiceAccountFilePath();
+    SparkGCPCredentials credentials;
+    if (serviceAccountFilePath != null) {
+      // if credential file is provided then use it to build SparkGCPCredentials
+      LOG.debug("Building credentials from specified service account file.");
+      credentials = new Builder().jsonServiceAccount(serviceAccountFilePath).build();
+    } else {
+      // this will use the application default credential which is available when running on google cloud
+      LOG.debug("No service account file specified. Building credentials using Application Default Credentials.");
+      credentials = new Builder().build();
+    }
+
+    JavaReceiverInputDStream<SparkPubsubMessage> pubSubMessages =
+      PubsubUtils.createStream(streamingContext.getSparkStreamingContext(), config.getProject(), config.topic,
```

Comment:
what happens if the topic is not actually the topic associated with the subscription? Is it simply ignored because the subscription already exists?

Reply:
Yes.

Reply:
Yes.